app/src/main/java/org/mozilla/fenix/home/blocklist/BlocklistMiddleware.kt

```diff
@@ -0,0 +1,99 @@
+ * Additionally, it reacts to state changes in order to filter them by the blocklist.
+ *
+ * @param settings Blocklist is stored here as a string set
+ */
+class BlocklistMiddleware(
+    private val blocklistHandler: BlocklistHandler
+) : Middleware<HomeFragmentState, HomeFragmentAction> {
+
+    /**
+     * Will filter "Change" actions using the blocklist and use "Remove" actions to update
+     * the blocklist.
+     */
+    override fun invoke(
+        context: MiddlewareContext<HomeFragmentState, HomeFragmentAction>,
+        next: (HomeFragmentAction) -> Unit,
+        action: HomeFragmentAction
+    ) {
+        next(getUpdatedAction(context.state, action))
+    }
+
+    private fun getUpdatedAction(
+        state: HomeFragmentState,
+        action: HomeFragmentAction
+    ) = with(blocklistHandler) {
+        when (action) {
+            is HomeFragmentAction.Change -> {
+                action.copy(
+                    recentBookmarks = action.recentBookmarks.filteredByBlocklist(),
+                    recentTabs = action.recentTabs.filteredByBlocklist(),
+                    recentHistory = action.recentHistory.filteredByBlocklist()
+                )
+            }
+            is HomeFragmentAction.RecentTabsChange -> {
+                action.copy(
+                    recentTabs = action.recentTabs.filteredByBlocklist()
+                )
+            }
+            is HomeFragmentAction.RecentBookmarksChange -> {
+                action.copy(
+                    recentBookmarks = action.recentBookmarks.filteredByBlocklist()
+                )
+            }
+            is HomeFragmentAction.RecentHistoryChange -> {
+                action.copy(recentHistory = action.recentHistory.filteredByBlocklist())
+            }
+            is HomeFragmentAction.RemoveRecentTab -> {
+                if (action.recentTab is RecentTab.Tab) {
+                    addUrlToBlocklist(action.recentTab.state.content.url)
+                    state.toActionFilteringAllState(this)
+                } else {
+                    action
+                }
+            }
+            is HomeFragmentAction.RemoveRecentBookmark -> {
+                action.recentBookmark.url?.let { url ->
+                    addUrlToBlocklist(url)
+                    state.toActionFilteringAllState(this)
+                } ?: action
+            }
+            is HomeFragmentAction.RemoveRecentHistoryHighlight -> {
+                addUrlToBlocklist(action.highlightUrl)
+                state.toActionFilteringAllState(this)
+            }
+            else -> action
+        }
+    }
+
+    // When an item is removed from any part of the state, it should also be removed from any other
+    // relevant parts that contain it.
+    // This is a candidate for refactoring once context receivers lands in Kotlin 1.6.20
+    // https://blog.jetbrains.com/kotlin/2022/02/kotlin-1-6-20-m1-released/#prototype-of-context-receivers-for-kotlin-jvm
```

Comment:
Interesting! How do you envision this will look like / improve with context receivers?

Reply:
I might just be excited to try them out 😅 

Replacing the `BlocklistHandler` parameter with a `context(BlocklistHandler)` receiver would mean the explicit `this` arguments could be removed from the calls to the function above, since they are already within a `BlocklistHandler` scope

Reply:
Ultimately I just like defining nested scopes as part of a functional type, rather than trying to resolve different scope levels mentally or with explicit `this@<scope>` syntax.